src/main/java/uk/gov/pay/api/filter/ratelimit/RedisRateLimiter.java

```diff
@@ -0,0 +1,103 @@
+package uk.gov.pay.api.filter.ratelimit;
+
+import com.google.inject.Inject;
+import com.google.inject.OutOfScopeException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.JedisPool;
+
+import java.time.LocalDateTime;
+import java.time.temporal.ChronoField;
+
+public class RedisRateLimiter {
+    private static final Logger LOGGER = LoggerFactory.getLogger(RedisRateLimiter.class);
+
+    private JedisPool jedisPool;
+
+    @Inject
+    public RedisRateLimiter(JedisPool jedisPool) {
+        this.jedisPool = jedisPool;
+    }
+
+    /**
+     * This block needs to be synchronous. Each RateLimit object will be shared between requests
+     * from the same source (Service), so is not shared across all the requests.
+     *
+     * @throws RateLimitException
+     */
+    synchronized void checkRateOf(String key, int rate, int perMillis)
+            throws RedisRateLimitException, RateLimitException {
+
+        Long count = null;
+
+        try {
+            count=updateAllowance(key, perMillis);
+        }
+        catch(Exception e){
+            // Exception possible if redis is unavailable or perMillis is too low or too high        
+            throw new RedisRateLimitException();
+        }
+
+        if (count!=null && count > rate) {
+            LOGGER.info(String.format("rate exceeded - count: %d, rate: %d", count, rate));
+            throw new RateLimitException();
+        }
+    }
+
+    synchronized private Long updateAllowance(String key, int perMillis) {
```

Comment:
consider renaming `perMillis` to `requestPerMillis`

Reply:
It is rate per millis.